File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == nullptr || head->next == nullptr || k == 0) {
            return head;
        }
        ListNode* temp = head;
        int len = 1;

        while(temp->next != nullptr) {
            temp = temp -> next;
            len++;
        }
        ListNode* tail = temp;
        tail->next = head;

        k = k % len;
        temp = head;
        int count = 0;
        int limit = len - k - 1;
        while(count < limit) {
            temp = temp -> next;
            count++;
        }

        ListNode* newTail = temp;
        ListNode* newHead = newTail -> next;
        newTail -> next = nullptr;

        return newHead;
    }
};

File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || k == 0) return head;
        int len = 1;
        ListNode tail = head;
        while(tail.next != null) {
            len++;
            tail = tail.next;
        }

        k = k % len;
        if(k == 0) return head;

        //Now we have Node tail at last
        tail.next = head;

        //now circular link is made
        //traverse till length - 1 node
        ListNode temp = head;
        int cnt = 1;
        while(temp != null) {
            if(cnt == (len - k)) {
                break;
            }
            cnt++;
            temp = temp.next;
        }

        //now we are at len - k node, update head and break link
        head = temp.next;
        temp.next = null;

        return head;
    }
}